src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -1,28 +1,115 @@
+                throw new ArgumentException(SR.Argument_MustBePrimitiveArray);
+
+            MethodTable* pMT = GetMethodTable(array);
+            nuint totalSize = pMT->ComponentSize * array.NativeLength;
+
+            uint size = RuntimeFieldHandle.GetFieldSize(new QCallFieldHandle(ref fldInfo));
+
+            // make certain you don't go off the end of the rva static
+            if (totalSize > size)
+                throw new ArgumentException(SR.Argument_BadFieldForInitializeArray);
+
+            void* src = (void*)RuntimeFieldHandle.GetStaticFieldAddress(fldInfo);
+            ref byte dst = ref MemoryMarshal.GetArrayDataReference(array);
+
+            Debug.Assert(!pMT->GetArrayElementTypeHandle().AsMethodTable()->ContainsGCPointers);
+
+            if (BitConverter.IsLittleEndian)
+            {
+                SpanHelpers.Memmove(ref dst, ref *(byte*)src, totalSize);
+            }
+            else
+            {
+                switch (pMT->ComponentSize)
+                {
+                    case 1:
+                        SpanHelpers.Memmove(ref dst, ref *(byte*)src, totalSize);
+                        break;
+                    case 2:
+                        BinaryPrimitives.ReverseEndianness(
+                            new ReadOnlySpan<ushort>(src, array.Length),
+                            new Span<ushort>(ref Unsafe.As<byte, ushort>(ref dst), array.Length));
+                        break;
+                    case 4:
+                        BinaryPrimitives.ReverseEndianness(
+                            new ReadOnlySpan<uint>(src, array.Length),
+                            new Span<uint>(ref Unsafe.As<byte, uint>(ref dst), array.Length));
+                        break;
+                    case 8:
+                        BinaryPrimitives.ReverseEndianness(
+                            new ReadOnlySpan<ulong>(src, array.Length),
+                            new Span<ulong>(ref Unsafe.As<byte, ulong>(ref dst), array.Length));
+                        break;
+                    default:
+                        Debug.Fail("Incorrect primitive type size!");
+                        break;
+                }
+            }
+        }
+
+        private static unsafe void* GetSpanDataFrom(
             RuntimeFieldHandle fldHandle,
             RuntimeTypeHandle targetTypeHandle,
-            out int count);
+            out int count)
+        {
+            IRuntimeFieldInfo fldInfo = fldHandle.GetRuntimeFieldInfo();
+
+            if ((RuntimeFieldHandle.GetAttributes(fldInfo.Value) & FieldAttributes.HasFieldRVA) == 0)
+                throw new ArgumentException(SR.Argument_BadFieldForInitializeArray);
+
+            TypeHandle th = new TypeHandle((void*)targetTypeHandle.Value);
+            Debug.Assert(!th.IsTypeDesc); // TypeDesc can't be used as generic parameter
+
+            if (!th.GetVerifierCorElementType().IsPrimitiveType()) // Enum is included
```

Comment:
This should not need to go through CorElementType. `MethodTable::IsPrimitive` should work fine.